### PR TITLE
Enable Non-Guild Gateways For requiredSettings

### LINKED ISSUE #435: Enable requiredSettings For Non-Guild Gateways
### Describe your proposal
Currently, `requiredSettings` property will only work for settings in the guild schema. There is no way to make sure that this user has this setting in the user or member settings. It is always assumed that the setting required to run the command is a guild setting.

### Use-Cases for your proposal
Add checks for settings from the member settings for member gateways.

### Expected and actual behavior
Currently, Klasa assumes that a setting that is required MUST be a guild setting. Non-guild settings will throw and cancel the commands usage if a member setting is added in when using the member gateway.

### Further details
Maybe if this wasn't an array of strings but tuples.
```js
requiredSettings: [
	['guild', 'settingName'], 
	['member', 'settingName'], 
	['author', 'settingName']
],
```
If the proposal is approved:

https://github.com/Skillz4Killz/klasa/pull/4/files

### PR DESCRIPTION
**Pull request by Skillz4Killz**
### Description of the PR
**BREAKING PR**

Closes #435 

### Changes Proposed in this Pull Request (List new items in CHANGELOG.MD)

- Converts `requiredSettings` from an array to an object that can hold arrays per gateway.

```js
requiredSettings: {
  guild: ['setting1', 'setting2'],
  member: ['setting3', 'setting4'],
  author: ['setting5', 'setting6']
}
```

### Suggestion

I would wait to merge this alongside another major breaking change to decrease the amount of breaking updates.

### Semver Classification

- [ ] This PR only includes documentation or non-code changes.
- [ ] This PR fixes a bug and does not change the (intended) framework interface.
- [ ] This PR adds methods or properties to the framework interface.
- [ ] This PR removes or renames methods or properties in the framework interface.

### CODE REVIEW

**Comment by UnseenFaith** [comment]
This PR should target the settings branch. We will compile a bunch of breaking changes into one PR at the end of it's development.

**Comment by Skillz4Killz** [comment]
Changed base to settings. Will review travis issues when i get a chance

**Comment by Skillz4Killz** [comment]
Travis issues fixed. Should be ready for review. If anything needs changing please let me know. :)

File: src/inhibitors/requiredSettings.js [hunk 1]
```diff
@@ -3,10 +3,19 @@ const { Inhibitor } = require('klasa');
...
-		if (!command.requiredSettings.length || message.channel.type !== 'text') return;
-		// eslint-disable-next-line eqeqeq
-		const requiredSettings = command.requiredSettings.filter(setting => message.guild.settings.get(setting) == null);
-		if (requiredSettings.length) throw message.language.get('INHIBITOR_REQUIRED_SETTINGS', requiredSettings);
+		const gateways = Object.keys(command.requiredSettings);
+		if (!gateways.length) return;
+
+		const missingSettings = [];
+
+		for (const gateway of gateways) {
```
Comment (KingDGrizzle) [inline src/inhibitors/requiredSettings.js]: Can't this use a for-of loop instead?

Comment (bdistin) [reply src/inhibitors/requiredSettings.js]: doesn't it already?

Comment (KingDGrizzle) [reply src/inhibitors/requiredSettings.js]: Woops, didn't word it right! I meant a for-of with Object.entries from the settings object

File: src/inhibitors/requiredSettings.js [hunk 2]
```diff
@@ -3,10 +3,17 @@ const { Inhibitor } = require('klasa');
...
 
 	async run(message, command) {
-		if (!command.requiredSettings.length || message.channel.type !== 'text') return;
-		// eslint-disable-next-line eqeqeq
-		const requiredSettings = command.requiredSettings.filter(setting => message.guild.settings.get(setting) == null);
-		if (requiredSettings.length) throw message.language.get('INHIBITOR_REQUIRED_SETTINGS', requiredSettings);
+		const missingSettings = [];
+
+		for (const [gateway, settings] of Object.entries(command.requiredSettings)) {
+			if (gateway === 'guild' && message.channel.type !== 'text') continue;
```
Comment (bdistin) [inline src/inhibitors/requiredSettings.js]: This is an opinion... If a command requires guild settings, then it should alert the user the settings don't exist just like any other missing settings.

File: src/inhibitors/requiredSettings.js [hunk 3]
```diff
@@ -3,10 +3,17 @@ const { Inhibitor } = require('klasa');
 module.exports = class extends Inhibitor {
 
 	async run(message, command) {
-		if (!command.requiredSettings.length || message.channel.type !== 'text') return;
-		// eslint-disable-next-line eqeqeq
-		const requiredSettings = command.requiredSettings.filter(setting => message.guild.settings.get(setting) == null);
-		if (requiredSettings.length) throw message.language.get('INHIBITOR_REQUIRED_SETTINGS', requiredSettings);
+		const missingSettings = [];
```
Comment (bdistin) [inline src/inhibitors/requiredSettings.js]: This is the wrong structure for storing settings from multiple locations. We should not be using tuples here. Object<gateway, Array\<settingName>>

File: src/languages/en-US.js [hunk 4]
```diff
@@ -57,7 +57,8 @@ module.exports = class extends Language {
 			INHIBITOR_MISSING_BOT_PERMS: (missing) => `Insufficient permissions, missing: **${missing}**`,
 			INHIBITOR_NSFW: 'You may not use NSFW commands in this channel.',
 			INHIBITOR_PERMISSIONS: 'You do not have permission to use this command.',
-			INHIBITOR_REQUIRED_SETTINGS: (settings) => `The guild is missing the **${settings.join(', ')}** guild setting${settings.length !== 1 ? 's' : ''} and thus the command cannot run.`,
+			// eslint-disable-next-line max-len
+			INHIBITOR_REQUIRED_SETTINGS: (settings) => settings.length === 1 ? `The ${settings[0][0]} is missing the ${settings[0][1]} setting.` : `The command was unable to be run because the following settings are missing:\n\n${settings.map(setting => `${setting[0]}: ${settings[1]}`)}`,
```
Comment (bdistin) [inline src/languages/en-US.js]: this should not list tuples, but rather:
```
guild: a, b, c, d
member: e, f, g
client: h
```
etc

**Comment by bdistin** [comment]
Changes were long since requested, and this pr is so far outdated, it has conflicts. I am closing this pr. If you would like to rewrite this pr targetting master and will make the changes in a timely manner as requested, you may do so. The idea isn't bad, but you have to work with us if you want to make a contribution.

**Pull request closed without merging**